content/posts/numpy/generator-rng/index.md

```diff
@@ -0,0 +1,304 @@
+
+Then `np.random.random` and `np.random.RandomState` calls 5 times. The global
+state is getting updated, but it does not impact our `rng` stream of numbers.
+
+A lot of packages already migrated their code to move away from the global
+generator in favour of an object one. In this case, setting the global state
+would have absolutely no effect and you will end up in a broken state.
+
+## Welcome to `np.random.Generator`
+
+Now you might have read that `np.random.RandomState` is deprecated and
+`np.random.Generator` should be used instead. It's true, but a bit more nuanced
+as we will see. What is true though is that in any case, a global state
+generator must not be used.
+
+So why this new interface `np.random.Generator`? The main reason is to break
+away from backward compatibilities concerns. `np.random.RandomState` was
+advertised as providing a constant bit-stream for a given seed. The problem
+is that this is extremely hard to guarantee for a wide range of
+platform/architecture. Even more if you want to change underlying methods,
+improve speed, fix bugs, etc. This is why NumPy decided with the policy
+described in
+[NEP19](https://numpy.org/neps/nep-0019-rng-policy.html)
+to create a new interface.
+
+> Did you know? `rand`, `randn`, etc. were just added to accommodate Matlab
+> users. These names match similar functions from Matlab to help code porting.
+
+Hence, for new code, `np.random.Generator` should be
+used. It does not provide a strong guarantee of reproducibility across
+different versions. But it's fine as we will see in the next section about
+[seed](#seed-id). To create a new instance of `np.random.Generator`, the
+canonical way is to use:
+
+```python
+rng = np.random.default_rng()
+```
+
+When you have to write tests `np.random.RandomState` is still recommended.
+Yes it's slower and has known issues, but it guarantees the bit-stream.
+Something you might need if you have sensitive tests.
+
+For more details have a look at the documentation of
+[`np.random`](https://numpy.org/doc/stable/reference/random/index.html)
+and at the policy document
+[NEP19](https://numpy.org/neps/nep-0019-rng-policy.html).
+
+## Goodbye seed {#seed-id}
+
+Imagine a forest. In a healthy ecosystem, the forest will consist of a multitude
+of species and each tree will be different. This collective group will work
+together to stay healthy. As a whole, a given forest will have a sort of
+homogeneity, a character. Now, burn everything down and use a single seed to
+rebuild your forest. This results in a monoculture, which is poor and less
+robust to the environment.
+
+Setting the seed of a random number generator is just like that. There is
+nothing random anymore, you are breaking your random number generator, breaking
+your forest. You see often in **code** the exact same seeds: 0, 123456.
+In fact, if you do this, you are using a fixed sequence of points which
+effectively has no guarantee over the quality of the points.
```

Comment:
Do we like funny half fitting references like this one: https://dilbert.com/strip/2001-10-25

The one tree we get is still just as random, but the forest will not look random anymore?  Maybe we should stress it a bit, that the tree is still pseudo random, but the forest may lose some of its randomness?